File: algorithms/java/204_count_primes/count_primes.java
/**
 * @author Peng
 *
 * Problem Description:
 *     Count the number of prime numbers less than a non-negative number, n.
 */
public class count_primes {

    /**
     * Solution 1: using 'Sieve of Eeatosthese' method
     * @return the number of prime number less than n
     */
    public static int count_primes(int n){
        if(n <= 1)  return 0;
        
        int count = 0;
        boolean [] flag = new boolean[n];  // mark flag to see whether the number has been marked to be 1
        for(int i = 2; i < n; i++) {
            if(!flag[i]) {
                count++;  // i is a prime
                for(int j = 2; i*j < n; ++j)  flag[i*j] = true;  // adjust the flag
            }
        }
        
        return count;
    }
    
    public static void main(String[] args) {
        System.out.println(count_primes(7));
    }

}

File: algorithms/python/204_count_primes/count_primes.py
# coding: utf8
# python 2.7
"""
Link: https://leetcode-cn.com/problems/count-primes
Count the number of prime numbers less than a non-negative number, n.
Example:
    Input: 10
    Output: 4
    Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.
"""
class Solution(object):
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        :simulate the composite number
        :complexity: time O(nlglgn), space: O(n)
        """
        flags = [1] * (n)  # init : true - prime, false - composite
        for i in range(2, n):
            if flags[i]:
                t = 2
                while i * t < n:
                    flags[i * t] = 0
                    t += 1
        return sum(flags[2:])

if __name__ == "__main__":
    so = Solution()
    res = so.countPrimes(10)
    print res
